### DISCUSSION TITLE
Is it possible to embed `deno_core` with a self-managed event loop?

### DISCUSSION THREAD

**Discussion by linrongbin16** [denoland/deno#25799] [Q&A] [answered]
And the 2nd question is: does the API `deno_core::JsRuntime::run_event_loop` only run 1 tick? or it try to wait everything inside its internal queue to complete? 

Because after reading some tutorials and source code of `deno_core`, I found there's already some queues inside `deno_core`, I guess they're for the `queueMicrotask` and built-in `async`/`await` functions.

For example, if we have below sample code:

```javascript

async function readFile(filename) {
   // Read a file in async mode
}

async function sleepFor(seconds) {
  // Sleep for seconds
}

const content = await readFile("README.md");
console.log(`We have read: ${content}`);
await sleepFor(10);
console.log(`We have sleep for 10s`);
```

Once the `deno_core::JsRuntime` execute it, I guess the 1st `console.log` will not be printed, instead it's been put in the internal queue and wait for the `async readFile` function done. Correct?

Then we call the `js_runtime.run_event_loop().await`  , once it's completed, is the only 1st `console.log` been printed? Or both two `console.log` are been printed? i.e. the `run_event_loop` API runs for only 1 tick, or wait for all the tasks done? 

If it only runs for 1 tick, then I can embed it into my command line TUI application. Because the TUI application uses `tokio` runtime with the `crossterm` library (it handles terminal events and renders terminal). I want to use `tokio::select!` on both the crossterm's `EventStream` (it receives terminal keyboard/mouse events, such as key pressed, mouse clicked, etc) and the `deno_core::JsRuntime::run_event_loop`.  Only 1 tick will never block the terminal events.

If it waits for all async tasks done, then it means it cannot work along with the tokio runtime managed by my TUI application. Because it can block the terminal.

> Note: I also post this question in discord: https://discord.com/channels/684898665143206084/1287318426909544528

**Comment by linrongbin16** [comment]
Oh, I found this: https://discord.com/channels/684898665143206084/1243892880188248084

Seems the `run_event_loop` will wait for all things done, which doesn't fit my requirement.
But the trick in the discord post is worth a try.

**Comment by bartlomieju** [comment, marked as answer]
> And the 2nd question is: does the API deno_core::JsRuntime::run_event_loop only run 1 tick? or it try to wait everything inside its internal queue to complete?

No, it runs for as long as event loop finishes all work. If you want to run a single "tick" you should use [`JsRuntime::poll_event_loop`](https://docs.rs/deno_core/latest/deno_core/struct.JsRuntime.html#method.poll_event_loop).

**Reply by linrongbin16** [reply]
hi, @bartlomieju thanks for reply!

Just one more question: I found the return value of `poll_event_loop` is a `Poll`.

I am a little confused how to wait for it done?

Or, it only returns `Ready` when done (just like `run_event_loop` is done), and for other cases it always returns `Pending`? correct?

**Reply by bartlomieju** [reply]
> Or, it only returns Ready when done (just like run_event_loop is done), and for other cases it always returns Pending? correct?

Yes, that's correct - if it returns `Poll::Pending` you should call `JsRuntime::poll_event_loop` some time in the future - in your case inside the `tokio::select!` macro.

**Reply by linrongbin16** [reply]
thanks again for your teaching/sharing! quite helpful.

**Comment by congzhangzh** [comment]
@bartlomieju hi, maybe this pattern possible? https://www.electronjs.org/blog/electron-internals-node-integration

**Comment by congzhangzh** [comment]
this maybe help https://github.com/congzhangzh/asyncio-guest/tree/master
